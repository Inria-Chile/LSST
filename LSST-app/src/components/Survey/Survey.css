
h2 {
  padding: 5px 10px;
}

h3 {
  background: #2d4053;
  padding: 5px 10px;
}

.survey-container {
  width: 100%;
}

.main-skymap-wrapper {
  background: #31475c;  
  /* background: #eee; */
  /* margin: 10px; */
  padding: 5px;
  position: relative;
}

.main-container {
  position: relative;
  width: 100%;
}

.left-container {
  position: absolute;
  top: 0;
  display: inline-block;
  width: 75%;
  margin: 0px 3px;
  padding: 0px 7px;
}

.right-container {
  position: absolute;
  top: 0;
  right: 0;
  display: inline-block;
  width: 25%;
  margin: 0px 3px;
  padding: 0px 7px;
}

.minimap {
  display: inline-block;
}

.minimap-container {
  right: 0;
  top: 0;
  background: #3d5871;   
}

.mainSkymap {
  display: block;
  /* margin-left: auto;
  margin-right: auto; */
  width: 44%;
  margin: 20px auto;
}

.survey-container {
  display: inline-block;  
}

.sidebar {
    display: none;
    height: 100%;
    width: 200px;
    position: fixed;
    text-align: left;
    padding: 15px;  
    z-index: 1;
    top: 0;
    right: -200px;
    background-color: #fff;
    border-left: 1px solid #111;
    overflow-x: hidden;
    transition: 0.5s;
}

.sidebar-open {
  width: 200px;
  right: 0px;
  padding: 15px;  
}

.settings-button {
  position: absolute;
  top: 0;
  right: 0;
  margin: 10px;
}

.hover-div {
  position: absolute;
  font-size: 13px;
  bottom: 0;
  left: 0;
}

.hover-filter {
  display: inline-block;
  margin: 0px 5px;
  text-align: center;
  border-radius: 2px;
  padding: 0px 5px;
}

/* #04070a */