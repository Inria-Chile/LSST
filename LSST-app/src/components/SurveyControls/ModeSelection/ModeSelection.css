
/* The switch - the box around the slider */
.switch {
  position: relative;
  display: inline-block;
  width: 60px;
  height: 14px;
  vertical-align: middle;
  margin: 0px 10px;
}

/* Hide default HTML checkbox */
.switch input {display:none;}

/* The slider */
.slider {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #aab1b9;
  -webkit-transition: .4s;
  transition: .4s;
}

.slider:before {
  position: absolute;
  content: "";
  height: 26px;
  width: 26px;
  left: 2px;
  bottom: -6px;
  background-color: #aab1b9;
  border: 5px solid white;
  -webkit-transition: .4s;
  transition: .4s;
}

input:checked + .slider {
  background-color: #aab1b9;
}

input:focus + .slider {
  box-shadow: 0 0 1px #aab1b9;
}

input:checked + .slider:before {
  -webkit-transform: translateX(32px);
  -ms-transform: translateX(32px);
  transform: translateX(32px);
}

/* Rounded sliders */
.slider.round {
  border-radius: 34px;
}

.slider.round:before {
  border-radius: 50%;
}

.highlight-text {
  color: white;
}