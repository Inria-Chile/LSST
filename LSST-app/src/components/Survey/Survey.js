import React, { Component } from 'react';
// import ReactDOM from 'react-dom';
import MainSkymap from '../Skymap/MainSkymap';
import MiniSkymaps from '../Skymap/MiniSkymaps';
import Charts from '../Charts/Charts';
import MiniScatterplot from '../Scatterplot/MiniScatterplot'
import MainScatterplot from '../Scatterplot/MainScatterplot'
import Sidebar from '../Sidebar/Sidebar';
import SurveyControls from '../SurveyControls/SurveyControls';
import ObservationsTable from '../ObservationsTable/ObservationsTable';
import './Survey.css';
import openSocket from 'socket.io-client';
import { filterColors, checkStatus, parseJSON } from "../Utils/Utils"

class Survey extends Component {
    
    static viewName = 'survey';

    constructor(props) {
        super(props);
        this.mainSkymap = null;
        this.miniSkymap = null;
        this.charts = null;
        this.miniScatterplot = null;
        this.mainScatterplot = null;
        this.displayedData = [];
        this.socket = openSocket(window.location.origin);
        this.state = {
            selectedMode: 'playback',
            selectedField: null,
            clickedField: [],
            displayedFilter: 'all',
            startDate: null,
            endDate: null,
            showSkyMap: true
        }
    }

    receiveMsg(msg){
        // console.log("received" + msg);
        msg.expDate = msg.request_time;
        this.addObservation(msg);
        this.setDate(new Date(parseInt(msg.request_time, 10)));
    }

    drawFrame = (timestamp) => {
        // console.log('Parent Drawframe')
        this.mainSkymap.drawFrame(timestamp);
        this.miniSkymap.drawFrame(timestamp);
        requestAnimationFrame(this.drawFrame);
    }
    
    setEcliptic = (show) => {
        this.mainSkymap.setEcliptic(show);
    }
    
    setGalactic = (show) => {
        this.mainSkymap.setGalactic(show);
    }
    
    setMoon = (show) => {
        this.mainSkymap.setMoon(show);
    }
    
    setTelescopeRange = (show) => {
        this.mainSkymap.setTelescopeRange(show);
    }
    
    setSidebar = (show) => {
        this.sidebar.setState({
            sidebarOpen: show
        })
    }
    
    setProjection = (proj) => {
        this.mainSkymap.setProjection(proj);
    }
    
    setDisplayedFilter = (filter) => {
        this.setState({
            displayedFilter: filter
        })
        if(this.mainSkymap!=null){
            this.mainSkymap.setDisplayedFilter(filter);
            
        }
        else{
            let showSkyMap = this.state.showSkyMap;
            this.setState({showSkyMap:!showSkyMap});
        }
       
    }

    displayScatterplot = () => {
        let showSkyMap = this.state.showSkyMap;
        this.setState({showSkyMap:!showSkyMap});
    }
    
    setDisplayedDateLimits = (startDate, endDate) => {
        if(this.state.showSkyMap)this.mainSkymap.setDisplayedDateLimits(startDate, endDate);
        this.miniSkymap.setDisplayedDateLimits(startDate, endDate);
        this.setDate(endDate);
        this.updateObservationsTable();
    }

    //function to set the start and end dates selected by the slider
    setTimeWindow = (startDate, endDate) => {
        this.startDisplayedDate = startDate;
        this.endDisplayedDate = endDate;
    }

    setLiveMode = () => {
        console.log('setlivemode')
        this.setState({
            selectedMode: 'live'
        })
        this.setData([]);
        console.log('SCOKERT', this.socket.on('data', timestamp => this.receiveMsg(timestamp)));
    }

    setPlaybackMode = () => {
        this.setState({
            selectedMode: 'playback'
        })
        this.socket.off('data');
        this.setData([]);
    }

    setDataByDate = (startDate, endDate) => {
        console.log('survey', 'Ssetdatabytdate')
        this.fetchDataByDate(startDate, endDate, (res) => {
            for(var i=0;i<res.results.length;++i)
                res.results[i]['fieldDec'] += 30;
            this.setData(res.results);
            this.setState({startDate: startDate, endDate: endDate});
            this.setDate(new Date(parseInt(endDate, 10)));
        })
    }

    setDate = (date) => {
        if(this.state.showSkyMap){
            this.mainSkymap.setDate(date);
        }
        this.miniSkymap.setDate(date);
    }
    
    fetchDataByDate = (startDate, endDate, cb) => {
        return fetch(`survey/playback/observationsCount?start_date=${startDate}&end_date=${endDate}`, {
            accept: "application/json"
        })
        .then(checkStatus)
        .then(parseJSON)
        .then(cb);
    }

    setData = (data) => {
        this.displayedData = data;
        this.charts.setData(data);
        this.miniScatterplot.setData(data);
        this.miniSkymap.setData(data);
        if(this.state.showSkyMap){
            this.mainSkymap.setData(data);
        }
        else this.mainScatterplot.setData(data);
    }

    addObservation = (obs) => {
        let added = false;
        for(let i=0;i<this.displayedData.length;++i){
            if(this.displayedData[i]['fieldID'] === obs['fieldID'] && this.displayedData[i]['filterName'] === obs['filterName']){
                // console.log('adding', obs, ' to ', this.displayedData[i]); // eslint-disable-line no-console          
                this.displayedData[i]['count']++;
                added = true;
            }
        }
        if(!added)
            this.displayedData.push(obs);
        this.setData(this.displayedData);
    }

    componentDidMount() {
        // console.log("componentDidMount")
        // this.fetchDataByDate(0, 99994323, (res) => {
        this.fetchDataByDate(0, 2, (res) => {
            for(var i=0;i<res.results.length;++i)
                res.results[i][2] += 30;
            // this.setData(res.results);
        });
    }

    cellHoverCallback = (fieldID, polygon) => {
        let latestField = null;
        let latestExpDate = 0;
        if(fieldID){
            for(let i=0;i<this.displayedData.length;++i){
                if(String(this.displayedData[i].fieldID) === String(fieldID) &&
                    (this.state.displayedFilter === this.displayedData[i].filterName || this.state.displayedFilter === 'all')){
                    if(this.displayedData[i].expDate > latestExpDate){
                        latestExpDate = this.displayedData[i].expDate;
                        latestField = this.displayedData[i];
                    }
                }
            }
        }
        this.setState({
            selectedField: latestField
        })
    }

    updateObservationsTable = () => {
        let fieldID = this.lastFieldID;
        let selectedFieldData = [];
        if(fieldID){
            for(let i=0;i<this.displayedData.length;++i){
                if(String(this.displayedData[i].fieldID) === String(fieldID) &&
                    (this.state.displayedFilter === this.displayedData[i].filterName || this.state.displayedFilter === 'all')){
                    selectedFieldData.push(this.displayedData[i]);
                }
            }
        }
        selectedFieldData.sort((a,b)=> b.expDate - a.expDate)
        this.setState({
            clickedField: selectedFieldData
        })
    }

    cellClickCallback = (fieldID, polygon) => {
        this.lastFieldID = fieldID;
        this.lastPolygon = polygon;
        this.updateObservationsTable();
        console.log('cellClickCallback');
    }

    componentDidUpdate(prevProps, prevState){
        if(prevState.displayedFilter !== this.state.displayedFilter)
            this.updateObservationsTable();
    }

    render() {
        let setters = {
            setEcliptic: this.setEcliptic,
            setGalactic: this.setGalactic,
            setMoon: this.setMoon,
            setTelescopeRange: this.setTelescopeRange,
            setSidebar: this.setSidebar,
            setProjection: this.setProjection
        }

        return (
            <div className="survey-container">
                <div>
                    <h2>
                        SURVEY PROGRESS MONITOR
                    </h2>
                    <button className="settings-button" type="button" onClick={this.setSidebar} aria-label="Settings">
                        <i className="fa fa-cog" aria-hidden="true"></i>
                    </button>
                </div>
                <div className="main-container">
                    <div className="left-container">
                        <SurveyControls setPlaybackMode={this.setPlaybackMode} 
                                        setLiveMode={this.setLiveMode} 
                                        setDataByDate={this.setDataByDate}
                                        selectedMode={this.state.selectedMode}
                                        startDate={this.state.startDate} 
                                        endDate={this.state.endDate} 
                                        setDisplayedDateLimits={this.setDisplayedDateLimits}/>
                        <Charts ref={instance => { this.charts = instance; }}/>
                        <div className="main-skymap-wrapper">
                            {!this.state.showSkyMap && <MainScatterplot ref={instance => {this.mainScatterplot=instance;}} data={this.displayedData} />}
                            {this.state.showSkyMap && <MainSkymap ref={instance => { this.mainSkymap = instance; }} 
                                data={this.displayedData} 
                                filter={this.state.displayedFilter}
                                startDate = {this.state.startDate} 
                                endDate={this.state.endDate}
                                cellHoverCallback={this.cellHoverCallback} 
                                cellClickCallback={this.cellClickCallback} />}
                            {
                                this.state.selectedField &&
                                <div className="hover-div">
                                    <div>FieldID: {this.state.selectedField && this.state.selectedField.fieldID ? this.state.selectedField.fieldID: ''}</div>
                                    <div>Timestamp: {this.state.selectedField && this.state.selectedField.expDate ? this.state.selectedField.expDate: ''}</div>
                                    <div>
                                        Filter: {" "}
                                        <div className="hover-filter" style={{
                                                backgroundColor: filterColors[this.state.selectedField.filterName ] ? filterColors[this.state.selectedField.filterName ]
                                                    : "#000000",
                                            }}>
                                            {
                                                this.state.selectedField.filterName ? this.state.selectedField.filterName : ''
                                            }
                                        </div>
                                    </div>
                                </div>
                            } 
                        </div>
                        <div>
                            <ObservationsTable selectedField={this.state.selectedField} clickedField={this.state.clickedField} />
                        </div>
                    </div>
                    <div className="right-container">
                        <MiniSkymaps ref={instance => { this.miniSkymap = instance; }} onMinimapClick={this.setDisplayedFilter} />
                        <MiniScatterplot ref={instance => {this.miniScatterplot=instance;}} onScatterplotClick={this.displayScatterplot}/>
                    </div>
                </div>
                <Sidebar ref={instance => { this.sidebar = instance; }} {...setters} skymap={this.mainSkymap} />
            </div>
        );
    }
}

export default Survey;
