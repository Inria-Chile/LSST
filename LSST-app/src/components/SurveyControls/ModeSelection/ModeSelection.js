import React, { Component } from 'react';
// import ReactDOM from 'react-dom';
import './ModeSelection.css';

class ModeSelection extends Component {

    constructor(props) {
        super(props);
        this.state = {
            playbackModeActive: true,
        }
    }

    handleChangeChk = () => {
        if(this.state.playbackModeActive)
            this.props.setLiveMode();        
        else
            this.props.setPlaybackMode();
        this.setState({
            playbackModeActive: !this.state.playbackModeActive,
        })
    }

    render() {
        return (
            <div className="mode-selection">
                <div>
                    <h6>Mode selection</h6>
                </div>
                <span className={this.state.playbackModeActive ? 'highlight-text' : ''}>Playback</span>
                <label className="switch">
                    <input type="checkbox" checked={!this.state.playbackModeActive} onChange={this.handleChangeChk}/>
                    <span className="slider round"></span>
                </label>
                <span className={!this.state.playbackModeActive ? 'highlight-text' : ''}>Live</span>
            </div>
        );
    }
}

export default ModeSelection;
