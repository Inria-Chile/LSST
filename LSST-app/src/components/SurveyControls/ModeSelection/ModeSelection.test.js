import React from 'react';
import ModeSelection from './ModeSelection';
import renderer from 'react-test-renderer';
import Survey from '../../Survey/Survey';

describe('ModeSelection_Test',function(){


//ASK: Se importo el suvey, esta bien asi? o habria que hacer algun mockup?.

    it('renders correctly without shallowRender', () => {
        const tree = renderer
          .create(<ModeSelection selectedMode={'playback'} 
            setPlaybackMode={Survey.setPlaybackMode} 
            setLiveMode={Survey.setLiveMode}/>)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });


    it('renders correctly without shallowRender', () => {
        const tree = renderer
          .create(<ModeSelection selectedMode={'live'} 
            setPlaybackMode={Survey.setPlaybackMode} 
            setLiveMode={Survey.setLiveMode}/>)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
})
