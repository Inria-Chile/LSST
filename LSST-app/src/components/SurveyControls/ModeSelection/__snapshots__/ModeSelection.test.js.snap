// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModeSelection_Test renders correctly without shallowRender 1`] = `
<div
  className="mode-selection"
>
  <div>
    <h6>
      Mode selection
    </h6>
  </div>
  <span
    className="highlight-text"
    onClick={[Function]}
  >
    Playback
  </span>
  <label
    className="switch"
  >
    <input
      checked={false}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="slider round"
    />
  </label>
  <span
    className=""
    onClick={[Function]}
  >
    Live
  </span>
</div>
`;

exports[`ModeSelection_Test renders correctly without shallowRender 2`] = `
<div
  className="mode-selection"
>
  <div>
    <h6>
      Mode selection
    </h6>
  </div>
  <span
    className=""
    onClick={[Function]}
  >
    Playback
  </span>
  <label
    className="switch"
  >
    <input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="slider round"
    />
  </label>
  <span
    className="highlight-text"
    onClick={[Function]}
  >
    Live
  </span>
</div>
`;
